man/man5/rlm_always.5
.\"     # DS - begin display
.de DS
.RS
.nf
.sp
..
.\"     # DE - end display
.de DE
.fi
.RE
.sp
..
.TH rlm_always 5 "10 January 2015" "" "FreeRADIUS Module"
.SH NAME
rlm_always \- FreeRADIUS Module
.SH DESCRIPTION
The \fIrlm_always\fP module provides a simple way to "always" return a
value during the processing of a configuration section.
.PP
The main configuration item is \fIrcode\fP, which sets the return code that
this instantiation of the module will return. The default, if none
specified, is 'fail'.
.PP
The valid options for rcode are as follows:
.RS
.TP
reject
reject the user;
.IP fail
a failure has occurred;
.IP ok
success;
.IP handled
the request has been handled: processing should be stopped and the response
sent;
.IP invalid
request is invalid;
.IP userlock
the user account has been locked out;
.IP notfound
the user account cannot be found;
.IP noop
no-op: nothing has happened;
.IP updated
the request has been updated.
.RE
.SH CONFIGURATION
.PP
.IP "rcode = <code>"
This module will always return with the code specified, as listed in the
table above.  If unspecified, the default is 'fail'.
.IP "simulcount = <n>"
If this module is used in the session{} section, the simulcount option
simulates the user having 'n' current sessions. The default is to not
override the number of sessions.
.IP "mpp = <yes|no>"
If set to yes, and this module is used in the session{} section, this
simulates the user having multilink sessions.  The default is 'no'.
.PP
.SH EXAMPLE
.PP
.DS
modules {
  ...
.br
  # instantiate the "always" module with the name "ok"
.br
  always ok {
.br
	# return code for this instantiation is "ok":
.br
	rcode = ok
.br
  }
.br
  ...
.br
}
.br

.br
recv Access-Request {
  ...
.br
  redundant {
     sql1          # try to find the user in sql1
.br
     sql2          # try to find the user in sql2
.br
     # the default here would be to fail, but...
.br
     ok            # if still not found, it's OK.
.br
  }
  ...
.br
}
.DE
.SH SECTIONS
.BR authorization,
.BR authentication,
.BR postauthentication,
.BR preaccounting,
.BR accounting,
.BR preproxy,
.BR postproxy
.PP
.SH FILES
.I /etc/raddb/mods-available/always
.PP
.SH "SEE ALSO"
.BR radiusd (8),
.BR radiusd.conf (5),
.BR unlang (5)
.PP
Further details of how module return codes operate can be found at <http://wiki.freeradius.org/config/Fail-over>.
.SH AUTHOR
Chris Parker <cparker@segv.org>, Matthew Newton
<matthew@newtoncomputing.co.uk>.


doc/ChangeLog
FreeRADIUS 3.0.15 Fri 26 May 2017 13:00:00 EDT urgency=medium
	Feature improvements

	Bug fixes
	* don't crash when reading corrupted data from session
	  resumption cache.  Fixes #1999.
	* Parse port in dhcpclient.  Fixes #2000.
	* Updated dictionary.starent

FreeRADIUS 3.0.14 Fri 26 May 2017 13:00:00 EDT urgency=medium
	Feature improvements
	* Enforce TLS client certificate expiration on
	  session resumption, and Session-Timeout.
	  See CVE-2017-9148.
	* Updated dictionary.cisco.vpn3000, dictionary.patton
	* Added dictionary.dellemc
	* Lowered the log output for failed PEAP sessions.
	* ALlow utc in rlm_date.  Patch from
	  Peter Lambrechtsen.
	* The internal OpenSSL session cache has been
	  disabled.  Please see mods-available/eap
	* Update detail reader documentation.
	  Patch from Matthew Newton.  Fixes #1973.
	* Make outgoing RadSec connections non-blocking.
	* Add SQL backing to Moonshot-*-TargetedId
	  generation.  Patch from Stefan Paetow.

	Bug fixes
	* radtest uses Cleartext-Password for EAP, not
	  User-Password.
	* Update documentation for mods-enabled/ linking.
	* Enhanced checks for moonshot salt.  Fixes #1933.
	* Allow session resumption for RadSec connections.
	  Fixes #1936.
	* Update "huntgroups" file to note that port ranges
	  are not supported.
	* Fix OpenSSL permissions issues on default key files.
	  Fixes #1941.
	* Certificates are not required when PSK is used.
	* Allow SubjectAltName as first extension in cert.
	  Fixes #1946.
	* Fixed talloc issue with TLS session resumption.
	  Fixes #1980.
	* "&Attr-26 := 0x01" now produces useful error messages.
	* Handle connection error in rlm_ldap_cacheable_groupobj.
	  Fixes #1951.
	* Fix endian issues in DHCP.
	* Multiple minor fixes for Coverity complaints.
	* Handle unexpected regex.  Fixes #1959.
	* Fix minor issues in dictionaries.
	* Fix typos and grammar.  Patches from Alan Buxey.
	* Fix erroneous VP creation in rlm_preproces.
	* Fix MIB.  Patch from Jeff Gehlbach.
	* Trust router updates from Alejandro Perez.
	* Allow build with LibreSSL.  Fixes #1989
	* Use correct packet for channel bindings.  Fixes #1990.
	* Many fixes found by PVS-Studio.  Thanks to PVS-Studio
	  for giving us a test license.  Please see the git commit
	  history for more information.
	* Fix incorrect length check in EAP-PWD.  This may
	  be exploitable.

FreeRADIUS 3.0.13 Mon 06 Mar 2017 13:00:00 EDT urgency=medium
	Feature improvements
	* Add dictionary.rfc7930.  Note that we do not implement
	  the RFC.
	* Added 'cipher_server_preference' to mods-available/eap
	  Patch from #1797.
	* OpenSSL 1.1.0 compatibility fixes.
	* rlm_perl: radiusd::xlat to evaluate xlat string
	  within perl script
	* Allow authentication retry in winbind. Patch from
	  Herwin Weststrate. See raddb/mods-available/mschap.
	* Added "recv-coa" method to rlm_rest.  It behaves the
	  same as "authorize".
	* Document Trust Router tr_port option.  Patch from
	  Stefan Paetow.
	* Update elasticsearch/logstash examples so that they work
	  with elastic stack v5.  Patch from Matthew Newton.
	* Print information about packets, replies, and contents
	  in the detail file reader.
	* Update abfab-tr policy.  Pull request #1893
	  from Stefan Paetow.
	* Reject packets which contain User-Password and
	  EAP-Message.
	* Add example for filtering Access-Challenge.
	  See sites-enabled/default.
	* Pull symlink fixes from v4.0.x.  Fixes #1859.
	* Add systemd reload.  Not everything is reloaded, but
	  some is.  Fixes #1662.
	* Better documentation for listen "ipaddr".  Fixes #1921
	* Add dictionary.cnergee, updated dictionary.nomadix.
	* radclient no longer needs -x to print statistics with -s.

	Bug fixes
	* Minor typos.  Fixes #1763
	* Fix typo in RPM build.  Closes #1767.
	* rlm_mschap check for password expiry only
	  if password was correct.  Fixes #1762.
	* Update debian build.
	* update rlm_counter "man" page.  Fixes #1775.
	* Remove erroneous assert.  Fixes #1778.
	* fix mschap password change test.  Fixes #1792.
	* Cleanup config file on data remove.  Fixes #1795.
	* passwd module returns "notfound" if not found.
	* Check for old OpenSSL, and don't build rlm_eap_fast
	  if it necessary.  Fixes #1803
	* Cleanup memory better after ldap version query.
	  Patch from Aleksey Katargin.
	* Rename lt_* functions to avoid linker issues with
	  libtool.  Fixes #1277
	* Many miscellaneous fixes and typos.
	* Allow long strings in %{%{foo} bar:-%{baz} blah".
	  Fixes #1866
	* Fix filtering operators, along with more documentation and
	  more tests for them.
	* Fix OpenSSL fixes.  Fixes #1876.
	* Finish SQL select queries even when SELECT returns no rows.
	  Fixes #1879.
	* Set Module-Failure-Message for more EAP errors.
	* Correct typo in dictionary.rfc5580.  Fixes #1882
	* Remove obselete systemd syslog.target.
	* Client-Port-Balance load-balancing now uses client port.
	* Radrelay examples fixed from Alex Clouter.
	* Update systemd target.  Pull request #1896.
	* Trim starting whitespace in xlat strings.
	* Get MySQL result lengths using normal API.
	* suid down after fchown().  Fixes #1914.
	* Fix cases of comparing pointer to NUL character.  Fixes #1915.
	* OpenSSL v1.1 fixes.  Pull request #1921.
	* Better Handle v4/v6 host names.  Pull request #1919.
	* Remove "Auth-Type = System" from docs and examples.
	* Don't crash on malformed %{home_server}.  Fixes #1922
	* fix erroneous use of talloc destructor in rlm_eap
	* Issue trigger modules.sql.fail.  Fixes #1923
	* Document python_path gotcha's.  Fixes #1845
	* dlopen() the specific version of Python.  Fixes #1592

FreeRADIUS 3.0.12 Thur 29 Sep 2016 13:00:00 EDT urgency=medium
	Feature improvements
	* Add support for =~ and !~ in update sections.
	  See "man unlang"
	* Add dictionary.checkpoint.
	* Simultaneous-Use prints out more information.
	* Print WARNING in debug mode when packets may be
	  truncated.
	* Added expansions %{home_server:state} and
	  %{home_server_pool:state}, which show the
	  state of the server / pool.

src/modules/rlm_always/rlm_always.c
/*
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 *
 */

/**
 * $Id$
 * @file rlm_always.c
 * @brief Return preconfigured fixed rcodes.
 *
 * @copyright 2000,2006  The FreeRADIUS server project
 */
RCSID("$Id$")

#define LOG_PREFIX "rlm_always (%s) - "
#define LOG_PREFIX_ARGS inst->name

#include <freeradius-devel/radiusd.h>
#include <freeradius-devel/modules.h>

/*
 *	The instance data for rlm_always is the list of fake values we are
 *	going to return.
 */
typedef struct rlm_always_t {
	char const	*name;		//!< Name of this instance of the always module.
	char const	*rcode_str;	//!< The base value.
	char const	*rcode_old;	//!< Make changing the rcode work with %{poke:} and radmin.

	rlm_rcode_t	rcode;		//!< The integer constant representing rcode_str.
	uint32_t	simulcount;
	bool		mpp;
} rlm_always_t;

/*
 *	A mapping of configuration file names to internal variables.
 */
static const CONF_PARSER module_config[] = {
	{ FR_CONF_OFFSET("rcode", FR_TYPE_STRING, rlm_always_t, rcode_str), .dflt = "fail" },
	{ FR_CONF_OFFSET("simulcount", FR_TYPE_UINT32, rlm_always_t, simulcount), .dflt = "0" },
	{ FR_CONF_OFFSET("mpp", FR_TYPE_BOOL, rlm_always_t, mpp), .dflt = "no" },
	CONF_PARSER_TERMINATOR
};

static int mod_instantiate(void *instance, CONF_SECTION *conf)
{
	rlm_always_t *inst = instance;

	inst->name = cf_section_name1(conf);
	if (!inst->name) inst->name = cf_section_name2(conf);
	/*
	 *	Convert the rcode string to an int
	 */
	inst->rcode = fr_str2int(mod_rcode_table, inst->rcode_str, RLM_MODULE_UNKNOWN);
	if (inst->rcode == RLM_MODULE_UNKNOWN) {
		cf_log_err(conf, "rcode value \"%s\" is invalid", inst->rcode_str);
		return -1;
	}
	inst->rcode_old = NULL;	/* Hack - forces the compiler not to optimise away rcode_old */

	return 0;
}

/** Reparse the rcode if it changed
 *
 * @note Look ma, no locks...
 *
 * @param inst Module instance.
 */
static void reparse_rcode(rlm_always_t *inst)
{
	rlm_rcode_t rcode;

	rcode = fr_str2int(mod_rcode_table, inst->rcode_str, RLM_MODULE_UNKNOWN);
	if (rcode == RLM_MODULE_UNKNOWN) {
		WARN("Ignoring rcode change.  rcode value \"%s\" is invalid ", inst->rcode_str);
		return;
	}

	inst->rcode = rcode;
	inst->rcode_old = inst->rcode_str;
}

/*
 *	Just return the rcode ... this function is autz, auth, acct, and
 *	preacct!
 */
static rlm_rcode_t CC_HINT(nonnull) mod_always_return(void *instance, UNUSED void *thread, UNUSED REQUEST *request)
{
	rlm_always_t *inst = instance;

	if (inst->rcode_old != inst->rcode_str) reparse_rcode(inst);

	return inst->rcode;
}

#ifdef WITH_SESSION_MGMT
/*
 *	checksimul fakes some other variables besides the rcode...
 */
static rlm_rcode_t CC_HINT(nonnull) mod_checksimul(void *instance, UNUSED void *thread, REQUEST *request)
{
	struct rlm_always_t *inst = instance;

	if (inst->rcode_old != inst->rcode_str) reparse_rcode(inst);

	request->simul_count = inst->simulcount;

	if (inst->mpp) request->simul_mpp = 2;

	return inst->rcode;
}
#endif

extern rad_module_t rlm_always;
rad_module_t rlm_always = {
	.magic		= RLM_MODULE_INIT,
	.name		= "always",
	.inst_size	= sizeof(rlm_always_t),
	.config		= module_config,
	.instantiate	= mod_instantiate,
	.methods = {
		[MOD_AUTHENTICATE]	= mod_always_return,
		[MOD_AUTHORIZE]		= mod_always_return,
		[MOD_PREACCT]		= mod_always_return,
		[MOD_ACCOUNTING]	= mod_always_return,
#ifdef WITH_SESSION_MGMT
		[MOD_SESSION]		= mod_checksimul,
#endif
		[MOD_PRE_PROXY]		= mod_always_return,
		[MOD_POST_PROXY]	= mod_always_return,
		[MOD_POST_AUTH]		= mod_always_return,
#ifdef WITH_COA
		[MOD_RECV_COA]		= mod_always_return,
		[MOD_SEND_COA]		= mod_always_return
#endif
	},
};
